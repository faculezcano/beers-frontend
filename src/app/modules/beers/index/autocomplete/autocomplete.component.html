<ng-autocomplete
  [data]="list"
  searchKeyword="value"
  (inputChanged)="input($event)"
  (selected)="handleSelected($event)"
  [itemTemplate]="itemTemp"
>
</ng-autocomplete>

<ng-template #itemTemp let-result>
  <div class="row gx-0 text-start">
    <div class="col p-2" [innerHTML]="result.value"></div>
    <div class="col-auto p-2 text-muted small d-flex align-items-end">
      {{ result.type }}
    </div>
  </div>
</ng-template>
