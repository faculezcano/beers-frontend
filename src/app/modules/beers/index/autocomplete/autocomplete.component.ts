import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { BeersService } from '../../../../services/beers.service';

@Component({
  selector: 'app-autocomplete',
  templateUrl: './autocomplete.component.html',
  styleUrls: ['./autocomplete.component.scss'],
})
export class AutocompleteComponent implements OnInit {
  @Input() url: string = '';
  @Output() selected = new EventEmitter<any>();

  public list: any[] = [];
  public value: string = '';

  private searchTimer: number | undefined;

  constructor(private beersService: BeersService) {}

  ngOnInit(): void {}

  input(term: string) {
    this.list = [];

    if (this.searchTimer) {
      clearTimeout(this.searchTimer);
    }

    if (term) {
      this.searchTimer = setTimeout(() => {
        this.search(term);
      }, 500);
    }
  }

  search(term: string) {
    this.beersService.search(term).subscribe((results) => {
      this.list = (results as any[]) || [];
    });
  }

  handleSelected(event: any) {
    this.selected.emit(event);
    this.value = event.value;
  }
}
