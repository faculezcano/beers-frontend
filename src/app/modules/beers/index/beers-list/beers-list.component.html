<div *ngIf="!loading">
  <app-list-item *ngFor="let beer of beers" [beer]="beer"></app-list-item>
</div>

<div *ngIf="loading" class="text-center">
  <span
    class="spinner-border text-muted"
    role="status"
    aria-hidden="true"
  ></span>
</div>
