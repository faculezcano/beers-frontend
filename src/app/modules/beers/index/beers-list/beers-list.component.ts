import { Component, Input, OnInit } from '@angular/core';
import { BeersService } from '../../../../services/beers.service';

@Component({
  selector: 'app-beers-list',
  templateUrl: './beers-list.component.html',
  styleUrls: ['./beers-list.component.scss'],
})
export class BeersListComponent implements OnInit {
  @Input() clearWhenNoParams: boolean = false;
  @Input() loadWithoutParams: boolean = true;
  @Input() loadOnInit: boolean = true;

  private _params: any = {};
  @Input() set params(params: any) {
    this._params = params;
    if (Object.keys(params).length != 0 || this.loadWithoutParams) {
      this.loadBeers();
    } else {
      if (this.clearWhenNoParams) {
        this.clear();
      }
    }
  }
  get params() {
    return this._params;
  }

  public beers: any[] = [];
  public loading = false;

  constructor(private beersService: BeersService) {}

  ngOnInit(): void {
    if (this.loadOnInit) {
      this.loadBeers();
    }
  }

  loadBeers() {
    this.loading = true;
    this.beersService.getBeers(this.params).subscribe((beers) => {
      this.beers = beers as any[];
      this.loading = false;
    });
  }

  clear() {
    this.beers = [];
  }
}
