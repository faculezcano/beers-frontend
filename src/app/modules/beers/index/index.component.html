<div class="container-fluid">
  <div class="row my-3">
    <div class="col"></div>
  </div>

  <div class="row">
    <div class="col-lg-3 col-md-6">
      <h3 class="text-truncate my-3">Top ingredients</h3>

      <app-top-ingredients></app-top-ingredients>
    </div>
    <div class="col-lg-6 my-3 order-1 order-lg-0">
      <h1>Beers</h1>

      <app-beers-list></app-beers-list>
    </div>

    <div class="col-lg-3 col-md-6 order-0 order-lg-1">
      <div class="row my-3">
        <div class="col">
          <h3>Top 10 Beers</h3>
        </div>

        <div class="col-sm-auto d-flex align-items-center justify-content-end">
          <app-autocomplete (selected)="search($event)"></app-autocomplete>
        </div>
      </div>

      <app-beers-list
        [loadOnInit]="false"
        [loadWithoutParams]="false"
        [params]="topBeersParams"
      ></app-beers-list>
    </div>
  </div>
</div>
