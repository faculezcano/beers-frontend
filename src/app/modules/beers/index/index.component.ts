import { Component, OnInit } from '@angular/core';
import { BeersService } from '../../../services/beers.service';
import { Beer } from '../../../models/beer';
import * as tinycolor from 'tinycolor2';

@Component({
  selector: 'app-beers',
  templateUrl: './index.component.html',
  styleUrls: ['./index.component.scss'],
})
export class IndexComponent implements OnInit {
  public beers: Array<Beer> = [];
  public loading = false;
  private searchTimer: number | undefined = undefined;

  constructor(private beersService: BeersService) {}

  ngOnInit(): void {
    this.loadBeers();
  }

  loadBeers(params = {}) {
    this.loading = true;

    this.beersService.getBeers(params).subscribe((beers: Array<Beer>) => {
      this.beers = beers;
      this.loading = false;
    });
  }

  // this function transforms srm value to color
  getSRMColor(srm: number | undefined): string {
    if (!srm) {
      let bodyStyles = window.getComputedStyle(document.body);
      return bodyStyles.getPropertyValue('--bs-secondary');
    }

    const color = new tinycolor('#F7D281');
    color.spin(-srm); // this gives reddish tone
    return color.darken(srm).toHexString();
  }

  searchInput(event: Event) {
    this.loading = true;

    const term = (<HTMLInputElement>event.target).value;
    if (this.searchTimer) {
      clearTimeout(this.searchTimer);
    }

    this.searchTimer = setTimeout(() => {
      this.search(term);
    }, 500);
  }

  search(term: string) {
    this.loadBeers({
      q: term,
    });
  }
}
