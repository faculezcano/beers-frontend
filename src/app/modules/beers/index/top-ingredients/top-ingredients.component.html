<div *ngFor="let ingredient of ingredients; let i = index" class="card mb-2">
  <div class="card-body">
    <div class="row gx-0">
      <div class="col-auto h2">
        {{ i + 1 }}
      </div>
      <div class="col text-end">
        {{ ingredient.name }}
        <div class="small text-muted">Used: {{ ingredient.count }}</div>
      </div>
    </div>
  </div>
</div>
