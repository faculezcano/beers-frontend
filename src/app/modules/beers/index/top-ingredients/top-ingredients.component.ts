import { Component, Input, OnInit } from '@angular/core';
import { BeersService } from '../../../../services/beers.service';

@Component({
  selector: 'app-top-ingredients',
  templateUrl: './top-ingredients.component.html',
  styleUrls: ['./top-ingredients.component.scss'],
})
export class TopIngredientsComponent implements OnInit {
  @Input() ingredients: any[] = [];

  constructor(private beersService: BeersService) {}

  ngOnInit(): void {
    this.beersService.getTopUsedIngredients().subscribe((ingredients) => {
      this.ingredients = ingredients as any[];
    });
  }
}
