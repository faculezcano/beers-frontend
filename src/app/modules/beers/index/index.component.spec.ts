import { ComponentFixture, TestBed } from '@angular/core/testing';

import { IndexComponent } from './index.component';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { BeersListComponent } from './beers-list/beers-list.component';
import { AutocompleteComponent } from './autocomplete/autocomplete.component';
import { AutocompleteLibModule } from 'angular-ng-autocomplete';
import { TopIngredientsComponent } from './top-ingredients/top-ingredients.component';

describe('BeersComponent', () => {
  let component: IndexComponent;
  let fixture: ComponentFixture<IndexComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [
        IndexComponent,
        BeersListComponent,
        AutocompleteComponent,
        TopIngredientsComponent,
      ],
      imports: [HttpClientTestingModule, AutocompleteLibModule],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(IndexComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
