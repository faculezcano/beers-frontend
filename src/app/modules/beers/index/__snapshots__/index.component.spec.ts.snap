// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeersComponent should match snapshot 1`] = `
<app-beers
  __ngContext__={[Function LRootView]}
  beersService={[Function BeersService]}
  topBeersParams={[Function Object]}
>
  <div
    class="container-fluid"
  >
    <div
      class="row my-3"
    >
      <div
        class="col"
      />
    </div>
    <div
      class="row"
    >
      <div
        class="col-lg-3 col-md-6"
      >
        <h3
          class="text-truncate my-3"
        >
          Top ingredients
        </h3>
        <app-top-ingredients>
        </app-top-ingredients>
      </div>
      <div
        class="col-lg-6 my-3 order-1 order-lg-0"
      >
        <h1>
          Beers
        </h1>
        <app-beers-list>
          <div
            class="text-center"
          >
            <span
              aria-hidden="true"
              class="spinner-border text-muted"
              role="status"
            />
          </div>
        </app-beers-list>
      </div>
      <div
        class="col-lg-3 col-md-6 order-0 order-lg-1"
      >
        <div
          class="row my-3"
        >
          <div
            class="col"
          >
            <h3>
              Top 10 Beers
            </h3>
          </div>
          <div
            class="col-sm-auto d-flex align-items-center justify-content-end"
          >
            <app-autocomplete>
              <div
                class="position-relative"
              >
                <ng-autocomplete
                  class="ng-autocomplete"
                  searchkeyword="value"
                >
                  <div
                    aria-expanded="false"
                    aria-owns="suggestions suggestions-history"
                    class="autocomplete-container"
                  >
                    <div
                      class="input-container"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-label=""
                        autocomplete="off"
                        class="ng-untouched ng-pristine ng-valid"
                        placeholder=""
                        role="combobox"
                        type="text"
                      />
                    </div>
                    <div
                      class="suggestions-container is-hidden"
                      id="suggestions"
                      role="listbox"
                    >
                      <ul>
                      </ul>
                    </div>
                    <div
                      class="suggestions-container is-visible"
                      id="suggestions-history"
                      role="listbox"
                    >
                      <ul>
                      </ul>
                    </div>
                  </div>
                </ng-autocomplete>
                <div
                  class="position-absolute h-100 end-0 top-0 d-flex align-items-center px-1"
                >
                  <i
                    class="bi-search mx-2 text-muted"
                  />
                </div>
              </div>
            </app-autocomplete>
          </div>
        </div>
        <app-beers-list>
          <div>
          </div>
        </app-beers-list>
      </div>
    </div>
  </div>
</app-beers>
`;
